server/bin/pbench-backup-tarballs.py

```diff
@@ -0,0 +1,220 @@
+#!/usr/bin/env python3
+# -*- mode: python -*-
+
+import os
+import sys
+import sh
+import glob
+import base64
+import signal
+import configtools
+import subprocess
+from argparse import ArgumentParser
+
+# Correct the first path to be the "lib" directory to make the
+# pbench-backup-to-s3 script simple, allowing the imports from the
+# backups3 library to work cleanly.
+first_path = sys.path[0]
+assert first_path.endswith('/bin')
+lib_path = os.path.join(os.path.dirname(first_path), 'lib')
+sys.path[0] = lib_path
+
+from pbench import PbenchConfig, BadConfig, get_pbench_logger
+from s3backup import S3Config
+
+_NAME_    = "pbench-backup-tarballs"
+
+# The link source and destination for this operation of this script.
+_linksrc = "TO-BACKUP"
+_linkdest = "BACKED-UP"
+
+# Global logger for the module, setup in main()
+_logger = None
+
+def rsync_archive_backup(config, report):
+    # Function to do rsync between archive and backup
+    exclude = ["--exclude={}".format(i) for i in config.EXCLUDE_DIRS.split(" ")]
+    # N.B The trailing slash is important
+    archive = "{}/".format(config.ARCHIVE)
+    stat_out = sh.rsync("-va", "--stats", exclude, archive, config.BACKUP)
+    if stat_out.exit_code == 0:
+        _logger.info(str(stat_out))
+        report.write(str(stat_out)[-384:])
+    elif stat_out.exit_code != 0:
+        _logger.warning("Failed to do rsync between {} and {}: {}\n".format(config.ARCHIVE, config.BACKUP, e))
+        report.write("Failed to do rsync between {} and {}: {}\n".format(
+            config.ARCHIVE, config.BACKUP, e))
+
+
+def _rename_tb_link(tb, dest, _logger):
+    try:
+        os.mkdir(dest)
+    except FileExistsError:
+        # directory already exists, ignore
+        pass
+    except Exception:
+        _logger.exception(
+            "os.mkdir: Unable to create tar ball destination directory: %s", dest)
+        raise
+    tbname = os.path.basename(tb)
+    tbnewname = os.path.join(dest, tbname)
+    try:
+        os.rename(tb, tbnewname)
+    except Exception:
+        _logger.exception(
+            "os.rename: Unable to move tar ball link %s to destination directory: %s", tb, dest)
+        raise
+
+
+def upload_to_bucket(s3_obj, tarlist, report, _logger):
+    try:
+        list_bucket_objects = s3_obj.connector.list_objects(
+            Bucket='{}'.format(s3_obj.bucket_name))
+    except Exception as e:
+        _logger.info(
+            "Bucket: {} is not present in s3\n{}".format(s3_obj.bucket_name, e))
+        return
+    try:
+        s3_contents = [s3_content_obj['Key']
+                        for s3_content_obj in list_bucket_objects['Contents']]
+    except Exception:
+        s3_contents = []
+    for tb in tarlist:
+        tar = os.path.realpath(tb)
+        filename = os.path.basename(tar)
+        if filename in s3_contents:
+            _logger.info(
+                "Result: {} already exists\n".format(filename))
```

Comment:
If the name already exists, but the MD5 is different what should we do?

Reply:
if someone tries to upload a duplicate object to s3, irrespective of object's md5sum, s3 overwrites the previous object with the new one. We should discuss this in details before proceeding.